### ISSUE TITLE
pkg-config references unknown shared libraries

### ISSUE THREAD

**Issue by sebastic** [Unidata/netcdf-c#1324]
The lintian QA tool reported several `pkg-config-references-unknown-shared-library` issues for the Debian package build of netcdf-c 4.6.2.1:

 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-lhdf5_hl` (line 11)
 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-lhdf5` (line 11)
 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-lpthread` (line 11)
 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-lsz` (line 11)
 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-lz` (line 11)
 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-ldl` (line 11)
 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-lm` (line 11)
 * `usr/lib/x86_64-linux-gnu/pkgconfig/netcdf.pc` `-lcurl` (line 11)

The lintian tag description is:
> The specified pkg-config(1) file references a shared object via, for
> example, Libs: -lfoo but this package appears to not ship the
> associated "libfoo.so" shared library.
> 
> This will result in a linker error and was likely caused by a missing
> installation step.
> 
> Please ensure that your package ships the corresponding libfoo.so
> shared object file.
> 
> Refer to the pkg-config(1) manual page and
> https://bugs.debian.org/919180 for details.

`netcdf.pc` contains:
```
prefix=/usr
exec_prefix=/usr
libdir=/usr/lib/x86_64-linux-gnu
includedir=/usr/include
ccompiler=/usr/bin/cc

Name: netCDF
Description: NetCDF Client Library for C
URL: http://www.unidata.ucar.edu/netcdf
Version: 4.6.2.1
Libs: -L${libdir} -L/usr/lib/x86_64-linux-gnu/hdf5/serial -lnetcdf -lhdf5_hl -lhdf5 -lpthread -lsz -lz -ldl -lm -lcurl
Cflags: -I${includedir} -I/usr/include/hdf5/serial
```

And `pkg-config(1)` documents:
> Libs:  This line should give the link flags specific to your package.  Don't add any flags for required packages; pkg-config will add those automatically.

`netcdf.pc.in` should probably not use `@NC_LIBS@` but only `-lnetcdf`.

### Environment Information

> Feel free to skip this if the issue is related to documentation, a feature request, or general discussion.

* What platform are you using? (please provide specific distribution/version in summary)
  * [x] Linux
  * [ ] Windows
  * [ ] OSX
  * [ ] Other
  * [ ] NA
* 32 and/or 64 bit?
  * [ ] 32-bit
  * [x] 64-bit
* What build system are you using?
  * [ ] autotools (`configure`)
  * [x] cmake
* Can you provide a sample netCDF file or `C` code to recreate the issue?
  * [ ] Yes (please attach to this issue, thank you!)
  * [x] No
  * [ ] Not at this time

**Comment by edhartnett** [comment]
If we only include -lnetcdf, that will work for shared library builds but not static-only builds.

We do not ship those libraries, they are created by other groups and we depend on them. If we take them out of the pkg-config file, then static build users will not know what libraries to link to in their code.

I am open to suggestions from anyone as to how to move forward on this in the most standard manner. I am not sure what should be done here.

**Comment by Dave-Allured** [comment]
Pkg-config has a standard solution for shared vs. static linking.  As @sebastic said, Libs: specifies only the link flags specific to the netcdf-C library.  Use Libs.private: for private libraries required by this package but not exposed to applications.  Suggest change `netcdf.pc.in` as follows.  This should be correct for both Mac and Linux, and both autotools and Cmake builds:
```
< Libs: -L${libdir} @NC_LIBS@
---
> Libs: -L${libdir} -lnetcdf
> Libs.private: -lhdf5_hl -lhdf5 -lz -ldl -lm -lcurl
```
This gives the expected output `netcdf.pc` from the configure stage, for both autotools and cmake.  For example, here is the resulting `netcdf.pc` from my Cmake configure test on Mac OS:
```
prefix=/usr/local
exec_prefix=/usr/local
libdir=/usr/local/lib
includedir=/usr/local/include
ccompiler=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc

Name: netCDF
Description: NetCDF Client Library for C
URL: http://www.unidata.ucar.edu/netcdf
Version: 4.6.2
Libs: -L${libdir} -lnetcdf
Libs.private: -lhdf5_hl -lhdf5 -lz -ldl -lm -lcurl
Cflags: -I${includedir}
```
Now users and higher level packages can query the link flags needed for dynamic linking:
```
$ pkg-config --libs netcdf
```
And for static linking, add the --static option:
```
$ pkg-config --libs --static netcdf
```
[This guide to pkg-config is helpful.](https://people.freedesktop.org/~dbn/pkg-config-guide.html)

Note 1:  I am new to pkg-config, so please check my suggested change.  I do not understand the difference between Libs.private and Requires.private.  Libs.private seemed sufficient in my initial testing.

Note 2:  The macro `@NC_LIBS@` was used originally to import all link flags in one chunk.  I don't know how to properly split this between the dynamic and static flags.  It would probably be best to generate the split as two separate macros from the originating configure script.

Note 3:  In current release 4.6.2, the macro `@NC_LIBS@` is generated differently between Autotools configure and Cmake configure.  This results in a key difference in the resulting `netcdf.pc`.  In my testing, Autotools configure emitted only the dynamic flag `-lnetcdf`, and NOT any of the static flags.  This means that the original reported problem appears only in Cmake builds, not Autotools.

**Comment by WardF** [comment]
Interesting solution, thanks for your input on this @Dave-Allured .  I'm preparing the 4.6.3 release; I still have a couple PR's to pull into it, I will try to make this change as well. Thanks!

**Comment by Dave-Allured** [comment]
Inclusion in 4.6.3 would be great.  This issue is bothering at least two repositories, Debian and Macports, in different ways.  Thanks.

**Comment by Dave-Allured** [comment]
Note 4:  Libs.private probably needs to be a macro, not hard coded like I showed initially.  This is because its contents probably change depending on netcdf's configure options.

**Comment by WardF** [comment]
The fix for this is in the branch `v4.6.3-dev-branch.wif`; this branch will be merged upstream into `master` today or tomorrow,. It has also propagated into the `v4.6.3-prep.wif` branch, which will be updated and tagged as the v4.6.3 release.  The tentative release date is tomorrow, February 28, 2019.
